doc/designs/hsm.md

```diff
@@ -0,0 +1,366 @@
+An HSM provides a hardened, tamper-resistant environment for secure
+cryptographic processing, key generation and encryption.
+
+The dogtagpki CA and KRA can utilize an HSM for private key
+generation and storage. 
+
+Access to the HSM is done using a PKCS#11-compliant shared library.
+
+The mechanism for sharing access to the key data for replicas/clones is HSM-specific and generally as a network accessible
+device.
+
+## Use Cases
+
+- Install IPA with a CA which has its private keys stored on an HSM self-signed
+- Install IPA with a CA which has its private keys stored on an HSM externally signed
+- Install an IPA replica CA which utilizes the same HSM-stored private keys
+- Install a KRA with private keys stored in an HSM
+- Install a KRA clone which utilizes the same HSM-stored private keys
+
+Installing with an HSM is all or nothing. There will be no mixing
+and matching the subsystems stored in the HSM. For example, one will
+not be able to install the KRA on an HSM without the CA as well.
+
+Only RSA keys will be supported.
+
+
+## How to Use
+
+Using an HSM should be largely invisible to users and administrators beyond passing additional options during installation. The options required and any pre-installion work are HSM-specific.
+
+It will not be possible to mix and match by default. PKI supports specifying the token values so a user can override these using --pki-config-override but it is, and will be, untested.
+
+There are a few basic rules:
+
+* If you use an HSM on the initial installation then all replicas and KRAs must also use the same HSM
+* You cannot upgrade an existing installation where the keys were not generated on an HSM to an HSM-based install.
+	+ It is likely that the HSM will not load externally-generated keys.
+	+ If you could it would require a lot of tedious effort to reconfigure the CA, certmonger tracking, the IPA CA LDAP entry and the NSS database to manage it.
+
+### Installation
+
+#### CA
+
+The token name, module name and shared library must be provided to the
+CA and KRA installers. These define where the keys will be stored
+and the naming convention for them.
+
+The CA signing, OCSP, Audit and Subsystem private keys and certificates
+are generated and stored in the HSM.
+
+| Option | Description                                       |
+|:------------------------ | :------------------------------ |
+| --token-name             | NSS name for the token          |
+| --library-path           | Path to PKCS#11 shared library  |
+| --token-password         | Password for the token          |
+
+This information will be stored in new schema so that replicas can auto-detect when an HSM is configured.
+
+ipa-ca-install will accept the same options.
+
+```
+attributeTypes: (
+  2.16.840.1.113730.3.8.21.1.TBD
+  NAME 'ipaCaHSMConfiguration'
+  DESC 'HSM Configuration'
+  SYNTAX 1.3.6.1.4.1.1466.115.121.1.15
+  SINGLE-VALUE
+  X-ORIGIN 'IPA v4.10'
+)
+```
+This will be added as a MAY to the `ipaCa` objectclass.
+
```

Comment:
The `ipaCa` objectclass is used by the IPA CA object but also to store the sub-cas, if any. The code will probably need to add checks to avoid setting inconsistent values for the token/path in the CA and sub CA.

2nd comment: the IPA CA object is replicated across all the replicas, for instance:
```
dn: cn=ipa,cn=cas,cn=ca,dc=ipa,dc=test
objectClass: top
objectClass: ipaca
cn: ipa
description: IPA CA
ipaCaId: cd487ec6-18c0-4dbd-ac75-b02c7eea70ec
ipaCaIssuerDN: CN=Certificate Authority,O=IPA.TEST
ipaCaSubjectDN: CN=Certificate Authority,O=IPA.TEST
ipaCaRandomSerialNumberVersion: 0
```
If the token name and library path are stored in this replicated entry, it means we enforce the same library path for all the CA instances. I'm wondering if there would be cases where replica1 has a different library path from replica2 because they have different architectures or OS.

Reply:
ok, I think I misinterpreted what you really meant. So it will be possible to run the following:
`ipa-server-install --token-name softhsm_token --library-path lib_path`
-> creates the ipaCaHSMConfiguration attribute: softhsm_token;lib_path and sets local PKI with `pki_hsm_libfile=lib_path`
Then for a replica with a different path: `ipa-replica-install --library-path other_lib_path`
-> does not update the ipaCaHSMConfiguration attribute but sets local PKI with `pki_hsm_libfile=other_lib_path`
And for an other replica re-using the same settings as the original master: `ipa-replica-install` (no specific option)
-> does not update the ipaCaHSMConfiguration attribute but reads the existing value and sets local PKI with `pki_hsm_libfile=value from LDAP`.

Reply:
Yes that was my intention.